pkg/scheduler/reconcile_on_location.go

```diff
@@ -19,33 +19,239 @@ package scheduler
+			logger.V(1).Info("begin to select Location", "edgePlacement", ep)
+			ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(ep)
+			if err != nil {
+				logger.Error(err, "invalid EdgePlacement key")
+				return err
+			}
+			currentSPS, err := c.singlePlacementSliceLister.Cluster(ws).Get(name)
+			if err != nil {
+				logger.Error(err, "failed to get SinglePlacementSlice", "workloadWorkspace", ws, "singlePlacementSlice", name)
+				return err
+			}
+
+			nextSPS := cleanSPSByLoc(currentSPS, lws.String(), lName)
+			nextSPS = extendSPS(nextSPS, singles)
+
+			_, err = c.edgeClusterClient.EdgeV1alpha1().SinglePlacementSlices().Cluster(ws.Path()).Update(ctx, nextSPS, metav1.UpdateOptions{})
+			if err != nil {
+				logger.Error(err, "failed to update SinglePlacementSlice", "workloadWorkspace", ws, "singlePlacementSlice", nextSPS.Name)
+			}
+		}
+	}
+
+	// dev-time tests
+	store.show()
+
 	return nil
 }
+
+// filterStsByLoc returns those SyncTargets that selected by the Location
+func filterStsByLoc(sts []*workloadv1alpha1.SyncTarget, loc *schedulingv1alpha1.Location) ([]*workloadv1alpha1.SyncTarget, error) {
+	filtered := []*workloadv1alpha1.SyncTarget{}
+	for _, st := range sts {
+		s := loc.Spec.InstanceSelector
+		selector, err := metav1.LabelSelectorAsSelector(s)
+		if err != nil {
+			return filtered, err
+		}
+		if selector.Matches(labels.Set(st.Labels)) {
+			filtered = append(filtered, st)
+		}
+	}
+	return filtered, nil
+}
+
+// filterEpsByLoc returns those EdgePlacements that select the Location
+func filterEpsByLoc(eps []*edgev1alpha1.EdgePlacement, loc *schedulingv1alpha1.Location) ([]*edgev1alpha1.EdgePlacement, error) {
+	filtered := []*edgev1alpha1.EdgePlacement{}
+	for _, ep := range eps {
+		for _, s := range ep.Spec.LocationSelectors {
+			selector, err := metav1.LabelSelectorAsSelector(&s)
+			if err != nil {
+				return filtered, err
+			}
+			if selector.Matches(labels.Set(loc.Labels)) {
+				filtered = append(filtered, ep)
+				break
+			}
+		}
+	}
+	return filtered, nil
+}
+
+// packEpKeys extracts keys from given EdgePlacements and put the keys in a map
+func packEpKeys(eps []*edgev1alpha1.EdgePlacement) map[string]empty {
+	keys := map[string]empty{}
+	for _, ep := range eps {
+		key, _ := kcpcache.MetaClusterNamespaceKeyFunc(ep)
+		keys[key] = empty{}
+	}
+	return keys
+}
+
+// cleanSPSByLoc removes all singleplacements that has the specified location, from a singleplacementslice
+func cleanSPSByLoc(sps *edgev1alpha1.SinglePlacementSlice, lws, lName string) *edgev1alpha1.SinglePlacementSlice {
+	nextDests := []edgev1alpha1.SinglePlacement{}
+	for _, sp := range sps.Destinations {
+		if sp.Cluster != lws || sp.LocationName != lName {
+			nextDests = append(nextDests, sp)
+		}
+	}
+	sps.Destinations = nextDests
+	return sps
+}
+
+func makeSinglePlacementsForLoc(locSelectingSts *schedulingv1alpha1.Location, sts []*workloadv1alpha1.SyncTarget) []edgev1alpha1.SinglePlacement {
```

Comment:
In this PoC there is supposed to be only one SyncTarget per Location.  We can not prevent the state from ever violating that restriction, but this function can treat it like the (hopefully transient) error condition that it is.